File: SenLib/Sen3/FileFixes/v0010_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class v0010_dat : FileMod {
		public string GetDescription() {
			return "Terminology fixes in Derfflinger Car 1.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4c2781204c47f245ul, 0x7e831b6a50413068ul, 0xb498c5b3u));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var patcher = new SenScriptPatcher(bin);

			// the Juno Naval Fortress -> Juno Naval Fortress
			patcher.RemovePartialCommand(0x26097, 0xe7, 0x26147, 0x4);
			patcher.RemovePartialCommand(0x262c6, 0x25, 0x262d0, 0x4);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/v0010.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4c2781204c47f245ul, 0x7e831b6a50413068ul, 0xb498c5b3u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/v0010.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/v0010_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "../file_fixes.h"

#include "../../p3a/pack.h"
#include "../../p3a/structs.h"
#include "../../sen/sen_script_patcher.h"
#include "../../sha1.h"

namespace SenLib::Sen3::FileFixes::v0010_dat {
std::string_view GetDescription() {
    return "Terminology fixes in Derfflinger Car 1.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/v0010.dat",
            293281,
            SenPatcher::SHA1FromHexString("4c2781204c47f2457e831b6a50413068b498c5b3"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;
        SenScriptPatcher patcher(bin);

        // the Juno Naval Fortress -> Juno Naval Fortress
        patcher.RemovePartialCommand(0x26097, 0xe7, 0x26147, 0x4);
        patcher.RemovePartialCommand(0x262c6, 0x25, 0x262d0, 0x4);

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::v0010_dat
